test/testComplexExchanger.js

```diff
@@ -0,0 +1,780 @@
+        it("check LOCKED", async function() {
+            let exchanger = await ComplexExchanger.deployed();
+            let deadline = + await exchanger.deadline.call();
+
+            jump = deadline - now();
+            console.log(now());
+            await timeMachine.jump(jump);
+            console.log(now());
+            let state = + await exchanger.getState.call();
+            //timeMachine.setTimestamp(now() - jump);
+            //await timeMachine.jump(-jump);
+            //await acc1.sendTransaction({from: acc2, value: web3.toWei(5,'ether')});
+            //reverter.revert();
+            console.log(now());
+            assert.equal(state, StateENUM.LOCKED,"Don't correct state!!");
+        });
+        */
+
+        it("(2) check PROCESSING_ORDERS", async function() {
+            let 
+                exchanger = await ComplexExchanger.deployed(),
+                state = + await exchanger.getState.call(),
+                calcTime = + await exchanger.calcTime.call();
+
+            if (now() - calcTime > exConfig.RATE_PERIOD) {
+                assert.notEqual(state, StateENUM.PROCESSING_ORDERS, 
+                    "Don't correct state when calcTime > RATE_PERIOD");
+                await exchanger.requestRates({value: web3.toWei(5,'ether')});
+                await exchanger.calcRates();
+                state = + await exchanger.getState.call();
+            }
+
+            assert.equal(state, StateENUM.PROCESSING_ORDERS,"Don't correct state when calcTime < RATE_PERIOD");
+        });
+
+        it("(3) check WAIT_ORACLES", async function() {
+            let exchanger = await ComplexExchanger.deployed();
+
+            await exchanger.requestRates({value: web3.toWei(5,'ether')});
+            let 
+                state = + await exchanger.getState.call(),
+                waiting = + await exchanger.waitingOracles.call();
+
+            if (waiting == 0) {
+                assert.notEqual(state, StateENUM.WAIT_ORACLES, "Don't correct state when waitingOracles == 0");
+                let oracle = await oracles[0].deployed();
+                await oracle.setWaitQuery(true);
+
+                state = + await exchanger.getState.call();
+            }
+
+            assert.equal(state, StateENUM.WAIT_ORACLES, "Don't correct state when waitingOracles != 0");
+        });
+
+        it("(4),(5) check CALC_RATES and REQUEST_RATES", async function() {
+            let exchanger = await ComplexExchanger.deployed();
+            await exchanger.requestRates({value: web3.toWei(5,'ether')});
+
+            let 
+                state = + await exchanger.getState.call(),
+                ready = + await exchanger.readyOracles.call();
+
+            if (ready >= exConfig.MIN_READY_ORACLES) {
+                assert.equal(state, StateENUM.CALC_RATES, 
+                    "Don't correct state when readyOracles >= MIN_READY_ORACLES");
+                for(let i=0; i < (oracles.length - exConfig.MIN_READY_ORACLES + 1); i++) {
```

Comment:
Check cycle condiftion. 
If we have 6 oracles, MIN_READY_ORACKES is 2:
6 - 2 + 1 = 5. 
Five is greater than 2, why we checks the state

Reply:
we check **state == CALC_RATES**, after that we broken 5 oracles, and check the **state != CALC_RATES**